From 13cf3cecd295087912ffd380c7c97a3e41495133 Mon Sep 17 00:00:00 2001
From: kancharx <kancharlax.suresh@intel.com>
Date: Wed, 3 Dec 2025 13:30:39 +0000
Subject: [PATCH] [CVE_2025_26436]Add equals method

Tracked-On: OAM-134642
Signed-off-by: kancharx <kancharlax.suresh@intel.com>

This adds an equals method which will for now only be used in tests.

Test: atest BackgroundStartPrivilegesTest
Flag: EXEMPT test only
Bug: 322159724
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:8aef21b72dca756458d25a42599779997d199f09)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:f6f2d41ffcd246afb758042b8fdbca9b5775c579)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:815cca137f84295c6037fd04e9de8011392a0ca6)
Merged-In: Ia726fa205b73a693fee775e4a08950ecd4b5f882
Change-Id: Ia726fa205b73a693fee775e4a08950ecd4b5f882
---
 .../app/BackgroundStartPrivileges.java        | 21 ++++++++++++++++++-
 .../app/BackgroundStartPrivilegesTest.java    | 11 ++++++++++
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/core/java/android/app/BackgroundStartPrivileges.java b/core/java/android/app/BackgroundStartPrivileges.java
index 20278eaee3b2..adea0a8a0702 100644
--- a/core/java/android/app/BackgroundStartPrivileges.java
+++ b/core/java/android/app/BackgroundStartPrivileges.java
@@ -23,12 +23,13 @@ import android.os.IBinder;
 import com.android.internal.util.Preconditions;
 
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Privileges granted to a Process that allows it to execute starts from the background.
  * @hide
  */
-public class BackgroundStartPrivileges {
+public final class BackgroundStartPrivileges {
     /** No privileges. */
     public static final BackgroundStartPrivileges NONE = new BackgroundStartPrivileges(
             false, false, null);
@@ -190,4 +191,22 @@ public class BackgroundStartPrivileges {
                 + ", originatingToken=" + mOriginatingToken
                 + ']';
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BackgroundStartPrivileges that = (BackgroundStartPrivileges) o;
+        return mAllowsBackgroundActivityStarts == that.mAllowsBackgroundActivityStarts
+                && mAllowsBackgroundForegroundServiceStarts
+                == that.mAllowsBackgroundForegroundServiceStarts
+                && Objects.equals(mOriginatingToken, that.mOriginatingToken);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(mAllowsBackgroundActivityStarts,
+                mAllowsBackgroundForegroundServiceStarts,
+                mOriginatingToken);
+    }
 }
diff --git a/core/tests/coretests/src/android/app/BackgroundStartPrivilegesTest.java b/core/tests/coretests/src/android/app/BackgroundStartPrivilegesTest.java
index cf6266c756ce..931d64640ea2 100644
--- a/core/tests/coretests/src/android/app/BackgroundStartPrivilegesTest.java
+++ b/core/tests/coretests/src/android/app/BackgroundStartPrivilegesTest.java
@@ -119,4 +119,15 @@ public class BackgroundStartPrivilegesTest {
                 Arrays.asList(BSP_ALLOW_A, BSP_ALLOW_A, BSP_ALLOW_A, BSP_ALLOW_A)))
                 .isEqualTo(BSP_ALLOW_A);
     }
+
+    @Test
+    public void backgroundStartPrivilege_equals_works() {
+        assertThat(NONE).isEqualTo(NONE);
+        assertThat(ALLOW_BAL).isEqualTo(ALLOW_BAL);
+        assertThat(ALLOW_FGS).isEqualTo(ALLOW_FGS);
+        assertThat(BSP_ALLOW_A).isEqualTo(BSP_ALLOW_A);
+        assertThat(NONE).isNotEqualTo(ALLOW_BAL);
+        assertThat(ALLOW_FGS).isNotEqualTo(ALLOW_BAL);
+        assertThat(BSP_ALLOW_A).isNotEqualTo(BSP_ALLOW_B);
+    }
 }
-- 
2.34.1

